services/debugger/cache/listcache/listCache.go

```diff
@@ -0,0 +1,107 @@
+package listcache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/config"
+)
+
+type cacheItem[E any] struct {
+	data       []E
+	lastAccess time.Time
+}
+
+/*
+loadCacheConfig sets the properties of the cache after reading it from the config file.
+This gives a feature of hot readability as well.
+*/
+func (c *ListCache[E]) loadCacheConfig() {
+	if c.size == 0 {
+		config.RegisterIntConfigVariable(3, &c.size, true, 1, "LiveEvent.cache.size")
+	}
+	if c.keyTTL == 0 {
+		config.RegisterDurationConfigVariable(3, &c.keyTTL, true, time.Hour, "LiveEvent.cache.ttl") // default keyTTL is 30 days
+	}
+	if c.cleanupFreq == 0 {
+		config.RegisterDurationConfigVariable(15, &c.cleanupFreq, true, time.Second, "LiveEvent.cache.clearFreq") // default clearFreq is 15 seconds
+	}
+}
+
+/*
+	Cache is an in-memory cache. Each key-value pair stored in this cache have a TTL and one goroutine removes the
+
+key-value pair form the cache which is older than TTL time.
+*/
+type ListCache[E any] struct {
```

Comment:
What do you think of using [this one](https://github.com/rudderlabs/rudder-server/blob/87a967e1a734789402db70d01b9e626699d40917/internal/throttling/cachettl/cachettl.go) instead? I designed it purposedly to avoid go routines and to increase performance. It's a double linked list with a hashmap to guarantee O(1) when getting values. The values are sorted in the double linked list by TTL to speed up the expiration process so that you don't have to iterate over all values each time (remember that we acquired a lock in order to clean up and that we don't know how many values we might have in the map. if they are sorted by ttl you can stop the moment you come across a value that has not expired). The double linked list is used also to speed up deletions and insertions.

```go
type ListCache[E any] struct {
	cache  cachettl.Cache[string, E]
	keyTTL time.Duration // Time after which the data will be expired and removed from the cache
}

/*
Update Inserts the data in the cache, This method expects a string as a key and []byte as the data
*/
func (c *ListCache[E]) Update(key string, value E) {
	c.cache.Put(key, value, c.keyTTL)
}
```

You might need to move it out of `internal` and add a `Pop` method. Let me know if it doesn't make sense and if you need assistance.

Reply:
Yeah I remember you using this in the throttling code but this has been present as legacy code for almost a year so I kept it intact but makes sense to re-use the one you implemented for redis

Reply:
I was looking into incorporating this but it seems like we don't have support to hold only a certain number of elements in memory. Let me know if I missed something

Reply:
We don't have it although it could be added along with a few tests. What would be the policy if the limit is reached? Refuse adding or eviction of older elements?

Reply:
@cisse21 if I understand correctly the plan is to remove `ListCache` once we're confident that badger works. please make sure to add a ticket for that.

Reply:
Added this [here](https://www.notion.so/rudderstacks/8aac9087df644365acdf64e28e290153?v=6e06b0a5ade24f0aa5ffe05dc2972e84&p=2500e96bbcc94e059b08ac2d28b795da&pm=s)